include/deal.II/base/exceptions.h

```diff
@@ -1438,6 +1438,8 @@ namespace deal_II_exceptions
                          const char       *exc_name,
                          ExceptionType     e)
     {
+      static_assert(std::is_base_of_v<ExceptionBase, ExceptionType>,
+                    "The provided exception must inherit from ExceptionBase.");
```

Comment:
Could the function just take a `const ExceptionBase &` as a reference?

Reply:
We call `e.set_fields()` so we might be able to pass this by reference, but definitely not by const reference.

Reply:
Sure, as a non-`const` reference is fine too.

It does not matter very much, but if you require that it's derived from the base class, you might as well express that through the type system rather than a `static_assert`.

Reply:
Hmm, good point. In some previous PRs I thought we did it this way (pass by-value) for some specific reason - I'll see if I can remember why.

Reply:
Would an rvalue reference `ExceptionBase &&` work?

Reply:
I thought it would, but it does not. Check this out:

https://github.com/dealii/dealii/blob/c04d7a36de5dc4e0ee59e408da8af8f2e55bf1cc/source/grid/tria.cc#L15652-L15673

Here we 'throw' a const object - that isn't going to work with rvalue references. I think this is the only place we do this but perhaps user code throws const objects too.

Reply:
Wow, wild! How does the call to `e.set_fields()` succeed in that case?

Reply:
This works because we make a non-const copy (for better or worse) in the function which actually throws the exception object.

Reply:
Ah, I see now!